### ISSUE TITLE
`ZipIterator` issues

### ISSUE THREAD

**Issue by elibarzilay** [Azure/mmlspark#72]
The current implementation of `ZipIterator` has a bunch of issues:

1. If it's possible, it would be much better to implement a plain
   iterator that returns the entries as lazy values that actually do the
   reading when needed.  This would make it possible to drop the
   sampling completely, and use something like
   `.filter(_ => r.nextDouble < someRatio)` instead of baking it in.

2. Reading the quick description, it's not clear to me that it always
   returns the same elements (ie, the `setSeed(0)`) -- but maybe this is
   idiomatic and shouldn't be documented?

3. Also, there is the known algorithm that returns N random elements,
   maybe it's also useful to do that?  (This would be easy if the first
   point is done.)

4. The implementation is not too great -- it looks like there are too
   many `var`s, and the `return` inside the `while` loop is making it
   hard to follow.  Again, doing the first point would make all of this
   complexity go away.

**Comment by mhamilton723** [comment]
This is no longer in the code as part of the streaming functionality pr
